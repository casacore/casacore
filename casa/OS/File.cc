//# File.cc: Class to define a File
//# Copyright (C) 1993,1994,1995,1996,1997,2001,2003
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casa/OS/Path.h>
#include <casa/OS/File.h>
#include <casa/OS/SymLink.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions.h>

#include <unistd.h>                 // needed for access, etc.
#include <sys/stat.h>               // needed for lstat or lstat64
#include <utime.h>                  // needed for utimbuf
#include <errno.h>                  // needed for errno
#include <casa/string.h>                 // needed for strerror
#include <casa/stdio.h>                  // needed for sprintf
#include <time.h>                   // needed for asctime/localtime on linux


namespace casa { //# NAMESPACE CASA - BEGIN

//# The ifdef's below are similar to those in IO/LargeIOFuncDef.h.
#if !defined(AIPS_NOLARGEFILE)
# ifdef AIPS_LINUX
#  if !defined(_LARGEFILE64_SOURCE)
#   define _LARGEFILE64_SOURCE
#  endif
# endif
#if defined(AIPS_DARWIN) || defined(AIPS_BSD)
# define fileLSTAT lstat
# define fileSTAT  stat
#else
# define fileLSTAT lstat64
# define fileSTAT  stat64
#endif
#else
# define fileLSTAT lstat
# define fileSTAT  stat
#endif


uInt File::uniqueSeqnr_p = 0;      // Initialization
Mutex File::theirMutex;


File::File () 
{
    // Sets itsPath on the current working directory
}

File::File (const Path& path) 
: itsPath (path)
{}

File::File (const String& string)
: itsPath (string)
{}

File::File (const File& that)
: itsPath (that.itsPath)
{}

File::~File()
{}
    
File& File::operator= (const File& that)
{
    if (this != &that) {
	itsPath = that.itsPath;
    }
    return *this;
}

Bool File::isRegular (Bool followSymLink) const
{
    // The struct is filled in by mylstat, and S_ISREG checks buf
    // if the file is a regularfile.
    Path testPath = itsPath;
    if (isSymLink()) {
	if (! followSymLink) {
	    return False;
	}
	testPath = SymLink(itsPath).followSymLink();
    }
    struct fileSTAT buf;
    if (mylstat (testPath.expandedName().chars(), &buf) < 0) {
	return False;
    }
    return  (S_ISREG (buf.st_mode));
}

Bool File::isDirectory (Bool followSymLink) const
{
    // The struct is filled in by mylstat, and S_ISDIR checks buf
    // if the file is a directory.
    Path testPath = itsPath;
    if (isSymLink()) {
	if (! followSymLink) {
	    return False;
	}
	testPath = SymLink(itsPath).followSymLink();
    }
    struct fileSTAT buf;
    if (mylstat (testPath.expandedName().chars(), &buf) < 0) {
	return False;
    }
    return  (S_ISDIR (buf.st_mode));
}

Bool File::isSymLink() const
{
    // The struct is filled in by mylstat, and S_ISLNK checks buf
    // if the file is a symbolic link.
    struct fileSTAT buf;
    if (mylstat (itsPath.expandedName().chars(), &buf) < 0) {
	return False;
    }
    return  (S_ISLNK (buf.st_mode));
}

Bool File::isPipe() const
{
    // The struct is filled in by mylstat, and S_ISFIFO checks buf
    // if the file is a pipe.
    struct fileSTAT buf;
    getstat (&buf);
    return (S_ISFIFO (buf.st_mode)); 
}

Bool File::isCharacterSpecial() const
{
    // The struct is filled in by mylstat, and S_ISCHR checks buf
    // if the file is a characterspecialfile.
    struct fileSTAT buf;
    getstat (&buf);
    return (S_ISCHR (buf.st_mode));
}

Bool File::isBlockSpecial() const
{
    // The struct is filled in by mylstat, and S_ISBLK checks buf
    // if the file is a blokspecialfile.
    struct fileSTAT buf;
    getstat (&buf);
    return (S_ISBLK (buf.st_mode));
}

Bool File::isSocket() const
{
    // The struct is filled in by mylstat, and S_ISSOCK checks buf
    // if the file is a socket.
    struct fileSTAT buf;
    getstat (&buf);
    return (S_ISSOCK (buf.st_mode));
}

Bool File::exists() const
{
    // The function access always substitutes symlinks.
    // Therefore use lstat instead.
    struct fileSTAT buf;
    return  (mylstat((itsPath.expandedName()).chars(), &buf) == 0);
}

Bool File::isReadable() const
{
    // The function access checks if the file is readable.
    return (access ((itsPath.expandedName()).chars(), R_OK)==0);
}

Bool File::isWritable() const
{
    // The function access checks if the file is writable.
    return (access ((itsPath.expandedName()).chars(), W_OK)==0);
}

Bool File::isExecutable() const
{
    // The function access checks if the file is executable.
    return (access ((itsPath.expandedName()).chars(), X_OK)==0);
}

Bool File::canCreate() const
{
    // Checks if the dirname of a file is a writable and executable
    // directory.
    File dir(itsPath.dirName());
    if (dir.isDirectory()  &&  dir.isWritable()  &&  dir.isExecutable()) {
	return True;
    }
    return False;
}

long File::userID() const
{
    // Returns the userid of a file which is extracted from the struct
    // buf.
    struct fileSTAT buf;
    getstat (&buf);
    return buf.st_uid;
}

long File::groupID() const
{
    // Returns the groupid of a file which is extracted from the struct
    // buf.
    struct fileSTAT buf;
    getstat (&buf);
    return buf.st_gid;
}

Int64 File::size() const
{
    // The struct buf is filled in by mylstat, and the size 
    // of the file is extracted from buf.
    struct fileSTAT buf;
    getstat (&buf);
    return buf.st_size;
}
    
uInt File::readPermissions() const
{
    // Returns the permissions as a decimal value. The value 
    // is extracted from buf.
    struct fileSTAT buf;
    getstat (&buf);
    return (uInt (buf.st_mode & 07) +
           (uInt (buf.st_mode &070) >> 3) * 10 +
           (uInt (buf.st_mode &0700) >> 6) * 100 );
}

void File::setPermissions(uInt permissions)
{
    // Changes the permissions by using chmod, the value must be 
    // an octal value.
    chmod ((itsPath.expandedName()).chars(),long (permissions));
}


Path File::newUniqueName (const String& directory, const String& prefix)
{
    // create an new unique name 
    char str[32];
    // fill str with the pid and the unique number
    uInt seqnr;
    {
      ScopedMutexLock lock(theirMutex); 
      seqnr = uniqueSeqnr_p++;
    }
    sprintf (str, "%i_%i", Int(getpid()), seqnr);
    if (directory.empty()  ||  directory.lastchar() == '/') {
	return Path (directory + prefix + str);
    }
    // a slash is added when a directory is given and the last
    // character is a slash.
    return Path (directory + "/" + prefix + str);
}

Path File::newUniqueName (const String& directory)
{
    // Uses the function newUniqueName with an empty prefix
    return newUniqueName (directory, "");
}

void File::touch(uInt time)
{
    // Uses the function utime to set the access time and the
    // modification time. 
    utimbuf times;
    times.actime = time;
    times.modtime = time;
    if (isWritable()) {
	utime ((itsPath.expandedName()).chars(), &times);
    }
}

void File::touch()
{
    // Uses the function utime to set the access time and the
    // modification time on the current time. 
    if (isWritable()) {
	utime ((itsPath.expandedName()).chars(), 0); 
    }
}

uInt File::accessTime () const
{
    // The struct is filled in by mylstat, and the accesstime 
    // is returned.
    struct fileSTAT buf;
    getstat (&buf);
    return buf.st_atime;
}

String File::accessTimeString () const
{
    // The struct is filled in by mylstat, and the accesstime 
    // is returned as a string.
    struct fileSTAT buf;
    getstat (&buf);
    return String (asctime (localtime (&buf.st_atime)));

}

uInt File::modifyTime () const
{
    // The struct is filled in by mylstat, and the modificationtime 
    // is returned.
    struct fileSTAT buf;
    getstat (&buf);
    return buf.st_mtime;
}


String File::modifyTimeString () const
{
    // The struct is filled in by mylstat, and the modificationtime 
    // is returned as a string.    
    struct fileSTAT buf;
    getstat (&buf);
    return String (asctime (localtime (&buf.st_mtime)));
}

File::FileWriteStatus File::getWriteStatus() const
{
  if (exists()) {
    return (isWritable()  ?  OVERWRITABLE : NOT_OVERWRITABLE);
  }
  return (canCreate()  ?  CREATABLE : NOT_CREATABLE);
}


uInt File::statusChangeTime () const
{
    // The struct is filled in by mylstat, and the statusChangetime 
    // is returned.
    struct fileSTAT buf;
    getstat (&buf);
    return buf.st_ctime;
}

String File::statusChangeTimeString() const
{
    // The struct is filled in by mylstat, and the statusChangetime 
    // is return from buf as a string.    
    struct fileSTAT buf;
    getstat (&buf);
    return String (asctime (localtime (&buf.st_ctime)));
}


void File::removeSymLinks ()
{
    // Remove the chain of symlinks starting with this one.
    File next(*this);
    while (next.isSymLink()) {
	SymLink symLink(next);
	next = symLink.readSymLink();
	symLink.remove();
    }
}

int File::mylstat(const char* path, void* buf) const
{
    return fileLSTAT (const_cast<char*>(path),
		      static_cast<struct fileSTAT*>(buf));
}

void File::getstat (const File& file, void* buf) const
{
    if (mylstat (file.path().expandedName().chars(), buf) < 0) {
	throw (AipsError ("File::getstat error on " +
			  file.path().expandedName() +
			  ": " + strerror(errno)));
    }
}

void File::checkTarget (Path& targetName, Bool overwrite,
			Bool forDirectory) const
{
    // Determine the target directory and file.
    // When the target is a directory, the basename is copied from the source.
    // Otherwise the target directory is the dirname of the target.
    Path targetDir;
    File target(targetName);
    if (!forDirectory  &&  target.isDirectory()) {
	targetDir = targetName;
	targetName.append (path().baseName());
    } else {
	targetDir = targetName.dirName();
    }
    // Check if the target directory is writable.
    target = targetDir;
    if (! target.isWritable()) {
	throw (AipsError ("RegularFile::copy/move: target directory " +
			  targetDir.originalName() + " is not writable"));
    }
    // Check if the target file exists, is writable and should be overwritten.
    target = targetName;
    if (target.exists()) {
	if (! overwrite) {
	    throw (AipsError ("RegularFile::copy/move: target file " +
			      targetName.originalName() + " already exists"));
	}
	if (! target.isWritable()) {
	    throw (AipsError ("RegularFile::copy/move: target file " +
			      targetName.originalName() +
			      " already exists and is not writable"));
	}
    }
}

} //# NAMESPACE CASA - END

