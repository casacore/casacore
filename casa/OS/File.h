//# File.h: Class to get file information and a base for other file classes
//# Copyright (C) 1993,1994,1995,1996,2000,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_FILE_H
#define CASA_FILE_H

//# Includes
#include <casa/aips.h>
#include <casa/OS/Path.h>
#include <casa/BasicSL/String.h>


namespace casa { //# NAMESPACE CASA - BEGIN

// <summary> 
// Class to get file information and a base for other file classes.
// </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>

// <use visibility=export>

// <prerequisite> 
//    <li> Basic knowledge of the UNIX file system 
//    <li> <linkto class=Path>Path</linkto> 
// </prerequisite>

// <etymology> 
// 'File' is used in a traditional sense.
// </etymology>

// <synopsis> 
// The File class provides the primary functions needed by all kinds of 
// files (directories, regular files, symbolic links, named pipes etc.).
// These shared functions serve mostly to return information about a 
// particular file -- for instance, its type, its ownership, read, write
// and execute permissions, date of latest access and the path on secundary 
// storage associated with this file. Every file object has, by definition, 
// a <linkto class=Path>Path</linkto> object associated with it which
// defines the file name.
// <p>
// See also the derived classes
// <linkto class=RegularFile>RegularFile</linkto>,
// <linkto class=Directory>Directory</linkto>, and
// <linkto class=SymLink>SymLink</linkto>.
// <br>
// This class does not contain virtual functions, because a lot of functions 
// have different parameters, e.g. 'create' for RegularFile has one parameter
// and 'create' for SymLink has two parameters. 
//
// It handles large files correctly.
// </synopsis>

// <example>
// <srcblock>
//    File myFile("someFileName");
//    if (myFile.exists()) {
//	myFile.setPermissions(0644);
//	if (myFile.isRegular()) {
//	    cout << "this file is a regular file" << endl;
//      }
//    }
//    else if (!myFile.exists()) {
//	  if (!myFile.canCreate()){
//	      cout << "cannot create this file" << endl;
//	  } 
//    }
// </srcblock>
// </example>

// <motivation> 
// File systems operations are a notorious source of porting problems.
// The file class provides a standard interface for programmers to use.
// </motivation>


class File
{
public: 

	enum FileWriteStatus {
		// file exists and can be overwritten
		OVERWRITABLE,
		// file exists but cannot be overwritten
		NOT_OVERWRITABLE,
		// file does not exist and is creatable
		CREATABLE,
		// file does not exist but cannot be created
		NOT_CREATABLE
	};


    // Construct a File object whose Path is set to the current working 
    // directory. 
    File();
    
    // Construct a File object whose Path is set to the given Path.
    // <group>
    File (const Path& path);
    File (const String& path);
    // </group>

    // Copy constructor (copy semantics).
    File (const File& that);

    virtual ~File();
    
    // Assignment (copy semantics).
    File& operator= (const File& that);

    // Returns the pathname of the file.
    const Path& path() const;

    // Check if the file is a regular file. If the boolean followSymLink is
    // False a symbolic link will not be followed.
    Bool isRegular (Bool followSymLink = True) const;

    // Check if the file is a directory. If the boolean followSymLink is
    // False a symbolic link will not be followed.
    Bool isDirectory (Bool followSymLink = True) const;

    // Check if the file is a symbolic link.
    Bool isSymLink() const;

    // Check if the file is a pipe.
    Bool isPipe() const;

    // Check if the file is a character special file.
    Bool isCharacterSpecial() const;

    // Check if the file is a block special file.
    Bool isBlockSpecial() const;

    // Check if the file is a socket.
    Bool isSocket() const;

    // Check if the file exists.
    Bool exists() const;

    // Check if the file is readable.
    Bool isReadable() const;

    // Check if the file is writable.
    Bool isWritable() const;

    // Check if the file is executable.
    Bool isExecutable() const;

    // Check if a file can be created.
    Bool canCreate() const;
    
    // Return the userID of the file.
    long userID() const; 

    // Return the groupID of the file.
    long groupID() const;
    
    // Return the size of the file. If the file
    // does not exist, an exception will be thrown.
    virtual Int64 size() const;

    // Return the permissions as a decimal value.
    uInt readPermissions() const;

    // Set permission with perm. Perm is an octal value.
    void setPermissions (uInt permissions);

    // Update access time and modification time of a file.
    void touch (uInt time);

    // Update access time and modification time of a file. This function
    // updates the file with the current time.
    void touch();

    // Time related fucnctions:
    // Return the time when the file was last accessed in seconds since
    // 00:00:00 GMT Jan 1, 1970.
    uInt accessTime() const;

    // Return the time when the file was last accessed
    // as a 26-characters String of the form:
    // Thu Feb  3 13:40:11 1994
    String accessTimeString() const;

    // Return the time when the file was last modified in seconds since
    // 00:00:00 GMT Jan 1, 1970.
    uInt modifyTime() const;

    // Return the time when the file was last modified
    // as a 26-characters String of the form:
    // Thu Feb  3 13:40:11 1994
    String modifyTimeString() const;

    // Return the time when the file status was last changed in seconds since
    // 00:00:00 GMT Jan 1, 1970.
    // It is set both by writing and changing the file status information,
    // such as changes of owner, group, link count, or mode.
    uInt statusChangeTime() const;

    // return the time when the file status was last changed
    // as a 26-characters String of the form:
    // Thu Feb  3 13:40:11 1994
    String statusChangeTimeString() const;

    // Create a new unique path name in the specified directory, with
    // the specified prefix and random trailing characters:
    // <srcblock>
    //    p.newUniqueName ("./", "temp")  -->  "./tempAAA00xx32"
    //    p.newUniqueName ("/home/me", "diary")  -->  "/home/me/diaryAAA00xxb0"
    // </srcblock>
    static Path newUniqueName (const String& directory, const String& prefix);

    // Create a new unique filename without a prefix.
    // As above, but all the characters in the filename are random:
    // <srcblock>
    //    p.newUniqueName ("./")  -->  "./AAA00xx32"
    //    p.newUniqueName ("/home/me")  -->  "/home/me/AAA00xxb0"
    // </srcblock>
    static Path newUniqueName (const String& directory); 


    // get write status of the file.
    // OVERWRITABLE - file exists and can be overwritten
    // NOT_OVERWRITABLE - file exists but cannot be overwritten
    // CREATABLE - File does not exist and can be created
    // NOT_CREATABLE - file does not exist and cannot be created.
    FileWriteStatus getWriteStatus() const;

protected:
    // This function is used by <linkto class=RegularFile>RegularFile</linkto> 
    // and <linkto class=Directory>Directory</linkto> to remove all the links
    // which, when followed, ultimately resolve to a Directory or a 
    // RegularFile.
    // For example, A->B, B->C, C->D and D points to a regular file.
    // When remove() is called for a regular file A,
    // that function uses removeLinks() to remove A, B, C and D.
    void removeSymLinks();

    // Check if the new path for a copy or move is valid.
    // An exception is thrown if:
    // <br>- the target directory is not writable
    // <br>- or the target file already exists and overwrite==False
    // <br>- or the target file already exists and is not writable
    // <br>When the targetName represents a directory, the basename
    // of the file is appended to it. This is done to cover the
    // case where the source is a symlink to a file. In that case
    // the target will get the basename of the symlink and not the
    // the basename of the file pointed to. This is not done when
    // forDirectory==True (which is used by class Directory).
    void checkTarget (Path& targetName, Bool overwrite,
		      Bool forDirectory = False) const;

private:
    // Define a function for lstat.
    // This is necessary since SunOS4.1.x prototypes lstat() with a first
    // argument of type (char*), while Solaris (and presumably all other
    // reasonable OS's) prototype it with a first argument of type
    // (const char*).  Since lstat() does not change its first argument,
    // it is safe to convert our const variable to a non-const one so that
    // we can call lstat() successfully.
    // <br>It is also useful to be able to pass the buffer as void*. In that
    // way the 32-bit or 64-bit file details are only needed in the cc file.
    int mylstat (const char* path, void* buf) const;

    // Get the lstat of this file.
    // Throw an exception when it fails.
    void getstat (void* buf) const;

    // Get the lstat of a file.
    // Throw an exception when it fails.
    void getstat (const File& file, void* buf) const;


    // A sequence number to generate unique file names.
    static uInt uniqueSeqnr_p;
    // Full pathname of the file.
    Path itsPath;
};


inline const Path& File::path() const
{
    return itsPath;
}

inline void File::getstat (void* buf) const
{
    getstat (*this, buf);
}




} //# NAMESPACE CASA - END

#endif
