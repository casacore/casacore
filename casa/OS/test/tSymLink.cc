//# tSymLink.cc: Test program for class SymLink
//# Copyright (C) 1993,1994,1995,1996,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casa/OS/SymLink.h>
#include <casa/OS/RegularFile.h>
#include <casa/OS/File.h>
#include <casa/BasicSL/String.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions.h>
#include <casa/iostream.h>


#include <casa/namespace.h>
// <summary>
// Test program for class SymLink.
// </summary>

// This program tests the class SymLink.
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.
// <p>
// When an argument is given, no exceptions will be thrown.
// This can be used to check if no memory leaks occur in normal operation.


void doIt (Bool doExcp)
{
    SymLink linkA ("tSymLink_tmp/A");
    SymLink linkI ("tSymLink_tmp/I");
    if (doExcp) {
	try {
	    SymLink symLink1(Path("tSymLink_tmp/isFile"));
	}
	catch (AipsError x) {                                   // regular file
	    cout << x.getMesg () << endl;
	} 
	try {
	    SymLink symLink1(Path("tSymLink_tmp/isDir"));
	}
	catch (AipsError x) {
	    cout << x.getMesg () << endl;                       // directory
	} 
    }

    SymLink newLink(Path("tSymLink_tmp/newLink"));
    SymLink newLink2(Path("tSymLink_tmp/isDir/newLink2"));
    File file("tSymLink_tmp/isDir");
    File file1("tSymLink_tmp/newLink");
    File file2("tSymLink_tmp/isDir/newLink2");
    file.setPermissions (0444);
    if (doExcp) {
	try {
	    SymLink symLink1(Path("tSymLink_tmp/isDir/newB"));
	}
	catch (AipsError x) {
	    cout << x.getMesg () << endl;                    // cannot create
	} 
	try {
	    newLink2.create("a");
	}
	catch (AipsError x) {
	    cout << x.getMesg () << endl;                    // cannot create
	} 
    }
    file.setPermissions (0755);
    AlwaysAssertExit (! file1.exists());
    AlwaysAssertExit (! file2.exists());

    // Create the new sym links.
    newLink.create ("s");
    newLink.create ("$HOME");
    newLink2.create ("../E");
    AlwaysAssertExit (file1.isSymLink());
    AlwaysAssertExit (file2.isSymLink());

    // Get the value of $HOME. If it is a symlink, follow it.
    File homeFile ("$HOME");
    String homeName;
    if (homeFile.isSymLink()) {
	homeName = SymLink(homeFile).followSymLink().originalName();
    }else{
	homeName = homeFile.path().expandedName();
    }

    // Read and follow some symlinks.
    AlwaysAssertExit (linkA.readSymLink().originalName() == "tSymLink_tmp/B");
    AlwaysAssertExit (linkI.readSymLink().originalName() == "tSymLink_tmp/H");
    AlwaysAssertExit (newLink.readSymLink().originalName() ==
		                              Path("$HOME").expandedName());
    AlwaysAssertExit (newLink2.readSymLink().originalName() ==
		                              "tSymLink_tmp/isDir/../E");
    if (doExcp) {
	try {
	    linkA.followSymLink();
	}
	catch (AipsError x) {
	    cout << x.getMesg () << endl;                       // endless loop
	} 
    }
    AlwaysAssertExit (linkI.followSymLink().originalName() == homeName);
    AlwaysAssertExit (newLink.followSymLink().originalName() == homeName);
    AlwaysAssertExit (newLink2.followSymLink().originalName() == homeName);

    // Recreate an existing symlink.
    linkI.create ("B");
    AlwaysAssertExit (linkI.readSymLink().originalName() == "tSymLink_tmp/B");
    if (doExcp) {
	try {
	    linkI.followSymLink();
	}
	catch (AipsError x) {
	    cout << x.getMesg () << endl;                       // endless loop
	} 
    }

    // Copy a symlink.
    SymLink linkII ("tSymLink_tmp/II");
    AlwaysAssertExit (! linkII.exists());
    linkI.copy ("tSymLink_tmp/II");
    AlwaysAssertExit (linkII.isSymLink());
    AlwaysAssertExit (linkI.isSymLink());
    AlwaysAssertExit (linkII.readSymLink().originalName() ==
		                                       "tSymLink_tmp/B");
    SymLink linkIII ("tSymLink_tmp/isDir/I");
    AlwaysAssertExit (! linkIII.exists());
    linkI.copy ("tSymLink_tmp/isDir");
    AlwaysAssertExit (linkIII.isSymLink());
    AlwaysAssertExit (linkI.isSymLink());
    AlwaysAssertExit (linkIII.readSymLink().originalName() ==
		                                       "tSymLink_tmp/isDir/B");

    // Move a symlink.
    File linkisFile ("tSymLink_tmp/isFile");
    AlwaysAssertExit (linkisFile.isRegular());
    linkI.move ("tSymLink_tmp/isFile");
    AlwaysAssertExit (linkisFile.isSymLink());
    AlwaysAssertExit (! linkI.exists());
    AlwaysAssertExit (SymLink(linkisFile).readSymLink().originalName() ==
                                                       "tSymLink_tmp/B");
    SymLink linkIIa ("tSymLink_tmp/isDir/II");
    AlwaysAssertExit (! linkIIa.exists());
    linkII.move ("tSymLink_tmp/isDir");
    AlwaysAssertExit (linkIIa.isSymLink());
    AlwaysAssertExit (! linkII.exists());
    AlwaysAssertExit (linkIIa.readSymLink().originalName() ==
		                                       "tSymLink_tmp/isDir/B");

    // Test symlink creation.
    SymLink test6;
    test6 = linkIIa;
    AlwaysAssertExit (test6.readSymLink().originalName() ==
		                                       "tSymLink_tmp/isDir/B");
    if (doExcp) {
	try {
	    test6.create ("a", False);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;               // already existing
	} 
    }
    SymLink test7("tSymLink_tmp/newLink2");
    RegularFile rfile("tSymLink_tmp/newLink2");
    rfile.create();
    if (doExcp) {
	try {
	    test7.create ("a");
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;               // already existing
	} 
    }
    rfile.remove();
    SymLink test8(test7);
    test8.create ("a", False);
    AlwaysAssertExit (test7.readSymLink().originalName() ==
		                                       "tSymLink_tmp/a");
}


int main (int argc, const char*[])
{
    try {
	doIt ( (argc<2));
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    cout << "OK" << endl;
    return 0;                           // exit with success status
}
