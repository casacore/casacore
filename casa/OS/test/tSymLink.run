#! /bin/sh
#-----------------------------------------------------------------------------
# Script to execute the tSymLink program via assay.
# It defines a few files, etc. to make the program work in any environment.
# Execution:
#      tSymLink.run        execute using assay (normal procedure)
#      tSymLink.run 0      only define the subdirectory, etc. (for debugging)
#      tSymLink.run 1      execute as such (to test)
#      tSymLink.run 1 1    execute without exceptions (to test leakage)
#=============================================================================

/bin/rm -rf tSymLink_tmp
mkdir tSymLink_tmp

ln -s  A  tSymLink_tmp/B
ln -s  B  tSymLink_tmp/A
ln -s  E  tSymLink_tmp/D
ln -s  $HOME  tSymLink_tmp/E
ln -s  E  tSymLink_tmp/F
ln -s  F  tSymLink_tmp/G
ln -s  G  tSymLink_tmp/H
ln -s  H  tSymLink_tmp/I
ln -s  $HOME  tSymLink_tmp/isLink
ls > tSymLink_tmp/isFile
mkdir tSymLink_tmp/isDir

if [ $# != 0 ]
then
    if [ $1 != 0 ]
    then
        $casa_checktool ./tSymLink $2
    fi
else
    $casa_checktool ./tSymLink
fi
