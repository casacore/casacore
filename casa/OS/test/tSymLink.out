SymLink: tSymLink_tmp/isFile exists, but is no symbolic link
SymLink: tSymLink_tmp/isDir exists, but is no symbolic link
SymLink: tSymLink_tmp/isDir/newB does not exist and cannot be created
SymLink::create error on a: Permission denied
SymLink: resolving tSymLink_tmp/A results in an endless loop
SymLink: resolving tSymLink_tmp/I results in an endless loop
SymLink::create: tSymLink_tmp/isDir/II already exists
SymLink::create: tSymLink_tmp/newLink2 already exists as a non-symlink
OK
